plugin/src/main/java/org/jenkinsci/plugins/workflow/cps/CpsFlowExecution.java

```diff
@@ -506,13 +508,53 @@ public FlowExecutionOwner getOwner() {
     private TimingFlowNodeStorage createStorage() throws IOException {
         FlowNodeStorage wrappedStorage;
 
-        FlowDurabilityHint hint = getDurabilityHint();
-        wrappedStorage = (hint.isPersistWithEveryStep())
-                ? new SimpleXStreamFlowNodeStorage(this, getStorageDir())
-                : new BulkFlowNodeStorage(this, getStorageDir());
+        if (this.storageDir != null && this.storageDir.endsWith("-completed")) {
+           wrappedStorage = new BulkFlowNodeStorage(this, getStorageDir());
+        } else {
+            FlowDurabilityHint hint = getDurabilityHint();
+            wrappedStorage = (hint.isPersistWithEveryStep())
+                    ? new SimpleXStreamFlowNodeStorage(this, getStorageDir())
+                    : new BulkFlowNodeStorage(this, getStorageDir());
+        }
         return new TimingFlowNodeStorage(wrappedStorage);
     }
 
+    /**
+     * Called when the build completes to migrate from {@link SimpleXStreamFlowNodeStorage} to
+     * {@link BulkFlowNodeStorage} to improve read performance for completed builds.
+     */
+    private void optimizeStorage(FlowNode flowEndNode) {
+        if (storage.delegate instanceof SimpleXStreamFlowNodeStorage) {
+            String newStorageDir = (this.storageDir != null) ? this.storageDir + "-completed" : "workflow-completed";
+            try {
+                FlowNodeStorage newStorage = new BulkFlowNodeStorage(this, new File(this.owner.getRootDir(), newStorageDir));
+                DepthFirstScanner scanner = new DepthFirstScanner();
+                scanner.setup(flowEndNode);
+                // The hope is that by doing this right when the execution completes, most of the nodes will already be
+                // cached in memory, saving us the cost of having to read them all from disk.
+                for (FlowNode node : scanner) {
+                    newStorage.storeNode(node, true);
+                }
+                newStorage.flush();
+                File oldStorageDir = getStorageDir();
+                synchronized (this) {
```

Comment:
Hmm, just noticed that `onProgramEnd` is already `synchronized`. The worst-case scan performance where many nodes have to be read from disk might be a problem. I guess we could trigger the migration here but run it asynchronously, and then give up on swapping `this.storage` below.

I would really like though for anything that only loads completed executions to only ever see the migrated storage.

Reply:
I think we keep this kind of approach. I marked the new method as `synchronized` just for clarity.